### DISCUSSION TITLE
ensure a shift cannot have two simillar shifts either side - shift_scheduling_sat.py

### DISCUSSION THREAD

**Discussion by braindeaf** [google/or-tools#2422] [CP-SAT questions] [answered]
Following the example of [shift_scheduling_sat.py](https://github.com/google/or-tools/blob/master/examples/python/shift_scheduling_sat.py), how could I create a constraint that ensures that a shift on the preceding day is present and/or a shift on the following day is present.

e.g.     
D L D D D
O L D D D
D L O D D
O L O D D <--- this is not valid as there is a day off either side of this L(ong) shift..

I had initially thought I could use similar logic to the penalized transitions, and it took me a while to work out how the penalized transitions blocked a transition with two .Not(). Which I understand now, but applying this to more than two variables gets awkward..

```python
    # Penalized transitions
    for previous_shift, next_shift, cost in penalized_transitions:
        for e in range(num_employees):
            for d in range(num_days - 1):
                transition = [
                    work[e, previous_shift, d].Not(), work[e, next_shift,
                                                           d + 1].Not()
                ]
                model.AddBoolOr(transition)
```

I appreciate it's a logic puzzle, and it's a question of saying....say

when work[2] is true
work[1] can be true or false
work[3] can be true or false
work[1] and work[3] cannot be false

I'm just struggling to turn that into code. Many thanks in advance :)

**Comment by lperron** [comment, marked as answer]
You can use implications:

```
D at day i imply O at day i + 1 is false (if the day exists)
D at day i imply O at day i - 1 is false (if the day exists)
```

**Reply by braindeaf** [reply]
I think maybe my question is just not very clear. I want to say that if

shift 2 is a L(ong) day, that is must have either a preceding or following D(ay) shift so the following would be valid

D L D
O L D
D L O

so it's something like

```
shifts = ['O', 'D', 'L', 'N']
days = [1,2,3,4,5]

model.addBoolOr([work[1, 1, 1], work[1, 1, 3]])  # if work[1, 3, 2]
```

so if the *if work[1,3,2]* should be an implication. What would that look like?

**Reply by lperron** [reply]
```
model.AddBoolOr([work[1, 3, 2].Not(), work[1, 1, 1], work[1, 1, 3]])
```

**Reply by braindeaf** [reply]
Thank you, this is just what I needed. I'm getting there slowly :)
